core/src/main/java/io/opencensus/stats/Measurement.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2016, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.stats;
+
+import com.google.auto.value.AutoValue;
+import io.opencensus.common.Function;
+import io.opencensus.stats.Measure.MeasureDouble;
+import io.opencensus.stats.Measure.MeasureLong;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Immutable representation of a Measure.
+ */
+@Immutable
+public abstract class Measurement {
+  /**
+   * Applies the given match function to the underlying data type.
+   */
+  public abstract <T> T match(Function<MeasurementDouble, T> p0, Function<MeasurementLong, T> p1);
```

Comment:
Oh sorry but this PR is built on top of https://github.com/census-instrumentation/opencensus-java/pull/405, so these parts will be updated in https://github.com/census-instrumentation/opencensus-java/pull/405. I'll rebase this PR once I've merged https://github.com/census-instrumentation/opencensus-java/pull/405, and for now please review  the last commit only.